Fix invalid references in monfaction code

Coolthulhu opened the pull request:
Monster faction adding code would get references to factions, add new factions and then use old references. If vector reallocation occurred when adding new factions, the references would become invalid.

BevapDin commented:
I solved that problem somewhere else by using pointers to the objects stored in a map (standard map, not unordered one).

The objects in the map are stable, adding entries does not affect existing ones. The vector only contains pointers into the map, the pointers stay valid all the time and can be resolved in constant time (no map look-up).

It has of course the disadvantage of required a standard (ordered) map.

Btw. how does that code behave when an (already loaded) faction is loaded again from a mod? Does it override the existing one, does it update/keep the references to it?

Another thing: don't worry much about what happens during the loading of the game data. The bottleneck there is definitively the json class, it does waaay to much parsing, especially repeated parsing of the same object. Making the storing faster (I refer to reserving space in the vector) will probably not make a difference. And if references become invalid, just use (copied) values, they won't be much worse and you avoid complicating the code.

Coolthulhu commented:
I only keep references to factions in that one block, so it should be safe now. It is relatively small, so complication isn't a big issue here.

> Btw. how does that code behave when an (already loaded) faction is loaded again from a mod? Does it override the existing one, does it update/keep the references to it?

Pretty sure it would modify the existing faction.
I wrote the code to be load order independent, because monster types need to be able to use factions that don't exist yet. In this case an empty faction is created with default values and then it is updated from json and later from parents and children.

> Making the storing faster (I refer to reserving space in the vector) will probably not make a difference.

I started writing the fix by reserving space. It alone could be enough to prevent the bug from happening, but I wanted to be sure so I added the explicit preallocation.
If simplicity is important here, the preallocation step could probably be scrapped if reserving can guarantee that the vector will not be reallocated after addition.

Though now that I changed the order of functions, the bug shouldn't happen even if I removed both the preallocation and reserving, because I don't mix adding factions and declaring references now.
